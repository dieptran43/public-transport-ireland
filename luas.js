'use strict'

/// <reference types="./dist/luas.d.ts" />
const luas = require('./dist/luas.js')

module.exports = luas
