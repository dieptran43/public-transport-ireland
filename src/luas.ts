/* eslint no-unused-vars: "off" */

import got from 'got'
import { parseXml } from './utils/parse-xml'

const SERVICE_URI = 'http://luasforecasts.rpa.ie/xml/get.ashx'

export enum Line {
  RED = 'Luas Red Line',
  GREEN = 'Luas Green Line'
}

interface AllStopsResponse {
  stops: {
    line: Array<{
      $: {
        name: Line // "Luas Red Line" or "Luas Green Line"
      }
      stop: Array<{
        _ : string // Stop name
        $ : {
          abrev: string
          isParkRide: string // "0" or "1"
          isCycleRide: string // "0" or "1"
          lat: string
          long: string
          pronunciation: string
          outboundStatusMessage: string
          outboundOperatingNormally: string // "True" or "False"
          outboundForecastsEnabled: string // "True" or "False"
          inboundStatusMessage: string
          inboundOperatingNormally: string // "True" or "False"
          inboundForecastsEnabled: string // "True" or "False"
        }
      }>
    }>
  }
}

interface Stop {
  /** The stop line (e.g "Luas Red Line") */
  line: Line
  /** The stop name (e.g "Saggart") */
  name: string
  /** The stop name pronunciation (e.g "Saggart") */
  pronunciation: string
  /** The stop abbreviated code (e.g "SAG") */
  code: string
  /** A boolean value indicating whether the current stop has a parking space */
  isParkRide: boolean
  /** A boolean value indicating whether the current stop has a cycle ride */
  isCycleRide: boolean
  /** The stop latitude (e.g. 53.28467885) */
  latitude: number
  /** The stop longitude (e.g. -6.43776255) */
  longitute: number
  /** Status message for outbound trains (e.g. "Services operating normally") */
  outboundStatusMessage: string
  /** boolean indicating whether outbound service is operating normally */
  outboundOperatingNormally: boolean
  /** boolean indicating whether outbound forecasts are enabled */
  outboundForecastsEnabled: boolean
  /** Status message for inbound trains (e.g. "Services operating normally") */
  inboundStatusMessage: string
  /** boolean indicating whether inbound service is operating normally */
  inboundOperatingNormally: boolean
  /** boolean indicating whether inbound forecasts are enabled */
  inboundForecastsEnabled: boolean
}

/**
 * Get all the available luas stop. Can filter by line
 */
export async function getStops (lineFilter?: Line) : Promise<Array<Stop>> {
  const response = await got(SERVICE_URI, {
    query: {
      action: 'list',
      ver: 2,
      encrypt: 'false' // yes, if you don't pass this one the response is encrypted -.-
    }
  })

  const stops : Array<Stop> = []
  const document : AllStopsResponse = await parseXml(response.body)
  document.stops.line.forEach((line) => {
    line.stop.forEach((stop) => {
      stops.push({
        line: line.$.name,
        name: stop._,
        pronunciation: stop.$.pronunciation,
        code: stop.$.abrev,
        isParkRide: stop.$.isParkRide === '1',
        isCycleRide: stop.$.isCycleRide === '1',
        latitude: Number(stop.$.lat),
        longitute: Number(stop.$.long),
        outboundStatusMessage: stop.$.outboundStatusMessage,
        outboundOperatingNormally: stop.$.outboundOperatingNormally === 'True',
        outboundForecastsEnabled: stop.$.outboundForecastsEnabled === 'True',
        inboundStatusMessage: stop.$.inboundStatusMessage,
        inboundOperatingNormally: stop.$.inboundOperatingNormally === 'True',
        inboundForecastsEnabled: stop.$.inboundForecastsEnabled === 'True'
      })
    })
  })

  if (lineFilter) {
    return stops.filter(stop => stop.line === lineFilter)
  }

  return stops
}

/**
 * get realtime information for a given stop
 */
export async function getRealTimeInfo (stop: string) {
  // TODO
  console.log(stop)
}

export default {
  getStops,
  getRealTimeInfo,
  Line
}
