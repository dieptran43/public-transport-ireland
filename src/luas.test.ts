import { readFileSync } from 'fs'
import { join } from 'path'
import nock from 'nock'
import luas from './luas'

const allStopsResponseData = readFileSync(join(__dirname, 'fixtures', 'luas', 'get-all-stops.xml'))
const realTimeInfoResponseData = readFileSync(join(__dirname, 'fixtures', 'luas', 'get-real-time-info.xml'))

// mock get stops data
nock('http://luasforecasts.rpa.ie')
  .get('/xml/get.ashx')
  .query({
    action: 'list',
    ver: '2',
    encrypt: 'false'
  })
  .twice()
  .reply(200, allStopsResponseData)

nock('http://luasforecasts.rpa.ie')
  .get('/xml/get.ashx')
  .query({
    action: 'forecast',
    stop: 'DOM',
    ver: '2',
    encrypt: 'false'
  })
  .times(3)
  .reply(200, realTimeInfoResponseData)

test('It should get all stops', async function () {
  const stops = await luas.getStops()
  expect(stops).toMatchSnapshot()
})

test('It should get all stops for a line', async function () {
  const greenLineStops = await luas.getStops(luas.Line.GREEN)
  expect(greenLineStops).toMatchSnapshot()
})

test('It should get real time information for a stop', async function () {
  const realTimeInfo = await luas.getRealTimeInfo('DOM')
  expect(realTimeInfo).toMatchSnapshot()
})

test('It should get real time information for a stop filtering by a given direction', async function () {
  const inboundRealTimeInfo = await luas.getRealTimeInfo('DOM', luas.Direction.INBOUND)
  expect(inboundRealTimeInfo).toMatchSnapshot()
})

test('It should get a stop status', async function () {
  const stopStatus = await luas.getStopStatus('DOM')
  expect(stopStatus).toMatchSnapshot()
})
