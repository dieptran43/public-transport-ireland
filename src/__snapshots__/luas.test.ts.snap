// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should get a stop status 1`] = `
Array [
  Object {
    "direction": "Inbound",
    "forecastsEnabled": true,
    "operatingNormally": true,
    "statusMessage": "Services operating normally",
  },
  Object {
    "direction": "Outbound",
    "forecastsEnabled": true,
    "operatingNormally": true,
    "statusMessage": "Services operating normally",
  },
]
`;

exports[`It should get all stops 1`] = `
Array [
  Object {
    "code": "TPT",
    "inboundForecastsEnabled": true,
    "inboundOperatingNormally": true,
    "inboundStatusMessage": "Services operating normally",
    "isCycleRide": false,
    "isParkRide": false,
    "latitude": 53.34835,
    "line": "Luas Red Line",
    "longitute": -6.22925833333333,
    "name": "The Point",
    "outboundForecastsEnabled": true,
    "outboundOperatingNormally": true,
    "outboundStatusMessage": "Services operating normally",
    "pronunciation": "The Point",
  },
  Object {
    "code": "SDK",
    "inboundForecastsEnabled": true,
    "inboundOperatingNormally": true,
    "inboundStatusMessage": "Services operating normally",
    "isCycleRide": false,
    "isParkRide": false,
    "latitude": 53.3488222222222,
    "line": "Luas Red Line",
    "longitute": -6.23714722222222,
    "name": "Spencer Dock",
    "outboundForecastsEnabled": true,
    "outboundOperatingNormally": true,
    "outboundStatusMessage": "Services operating normally",
    "pronunciation": "Spencer Dock",
  },
  Object {
    "code": "BRO",
    "inboundForecastsEnabled": true,
    "inboundOperatingNormally": true,
    "inboundStatusMessage": "Services operating normally",
    "isCycleRide": false,
    "isParkRide": false,
    "latitude": 53.37223956,
    "line": "Luas Green Line",
    "longitute": -6.29768465,
    "name": "Broombridge",
    "outboundForecastsEnabled": true,
    "outboundOperatingNormally": true,
    "outboundStatusMessage": "Services operating normally",
    "pronunciation": "Broombridge",
  },
  Object {
    "code": "CAB",
    "inboundForecastsEnabled": true,
    "inboundOperatingNormally": true,
    "inboundStatusMessage": "Services operating normally",
    "isCycleRide": false,
    "isParkRide": false,
    "latitude": 53.36433419,
    "line": "Luas Green Line",
    "longitute": -6.28196899,
    "name": "Cabra",
    "outboundForecastsEnabled": true,
    "outboundOperatingNormally": true,
    "outboundStatusMessage": "Services operating normally",
    "pronunciation": "Cabra",
  },
]
`;

exports[`It should get all stops for a line 1`] = `
Array [
  Object {
    "code": "BRO",
    "inboundForecastsEnabled": true,
    "inboundOperatingNormally": true,
    "inboundStatusMessage": "Services operating normally",
    "isCycleRide": false,
    "isParkRide": false,
    "latitude": 53.37223956,
    "line": "Luas Green Line",
    "longitute": -6.29768465,
    "name": "Broombridge",
    "outboundForecastsEnabled": true,
    "outboundOperatingNormally": true,
    "outboundStatusMessage": "Services operating normally",
    "pronunciation": "Broombridge",
  },
  Object {
    "code": "CAB",
    "inboundForecastsEnabled": true,
    "inboundOperatingNormally": true,
    "inboundStatusMessage": "Services operating normally",
    "isCycleRide": false,
    "isParkRide": false,
    "latitude": 53.36433419,
    "line": "Luas Green Line",
    "longitute": -6.28196899,
    "name": "Cabra",
    "outboundForecastsEnabled": true,
    "outboundOperatingNormally": true,
    "outboundStatusMessage": "Services operating normally",
    "pronunciation": "Cabra",
  },
]
`;

exports[`It should get real time information for a stop 1`] = `
Array [
  Object {
    "arrivingInMinutes": 9,
    "destination": "Broombridge",
    "direction": "Inbound",
    "expectedArrivalTime": "2019-09-23T17:57:19.000+01:00",
  },
  Object {
    "arrivingInMinutes": 17,
    "destination": "Broombridge",
    "direction": "Inbound",
    "expectedArrivalTime": "2019-09-23T18:05:19.000+01:00",
  },
  Object {
    "arrivingInMinutes": 0,
    "destination": "Sandyford",
    "direction": "Outbound",
    "expectedArrivalTime": "2019-09-23T17:48:19.000+01:00",
  },
  Object {
    "arrivingInMinutes": 9,
    "destination": "Sandyford",
    "direction": "Outbound",
    "expectedArrivalTime": "2019-09-23T17:57:19.000+01:00",
  },
  Object {
    "arrivingInMinutes": 17,
    "destination": "Sandyford",
    "direction": "Outbound",
    "expectedArrivalTime": "2019-09-23T18:05:19.000+01:00",
  },
]
`;

exports[`It should get real time information for a stop filtering by a given direction 1`] = `
Array [
  Object {
    "arrivingInMinutes": 9,
    "destination": "Broombridge",
    "direction": "Inbound",
    "expectedArrivalTime": "2019-09-23T17:57:19.000+01:00",
  },
  Object {
    "arrivingInMinutes": 17,
    "destination": "Broombridge",
    "direction": "Inbound",
    "expectedArrivalTime": "2019-09-23T18:05:19.000+01:00",
  },
]
`;
